src/rpc/mining.cpp

```diff
@@ -678,25 +682,50 @@ UniValue mkblocktemplate(const UniValue &params, int64_t coinbaseSize, CBlock *p
         // miners?
     }
 
+    const Consensus::Params &consensusParams = Params().GetConsensus();
+
     // Update block
     static CBlockIndex *pindexPrev = nullptr;
     static int64_t nStart = 0;
     static std::unique_ptr<CBlockTemplate> pblocktemplate(new CBlockTemplate());
-    if (pindexPrev != chainActive.Tip() || forceTemplateRecalc ||
-        (mempool.GetTransactionsUpdated() != nTransactionsUpdatedLast && GetTime() - nStart > 5))
+    static boost::shared_ptr<CReserveScript> prevCoinbaseScript;
+    static int64_t prevCoinbaseSize = -1;
+    // We cache the previous block templates returned, but we invalidate the
+    // cache below (generate a new block) if any of:
+    // 1. Global forceTemplateRecalc is true.
+    // 2. Cached block points to a different chaintip.
+    // 3. Is testnet and 30 seconds have elapsed (so we pick up the testnet
+    //    minimum difficulty -> 1.0 after 20 mins).
+    // 4. Mempool has changed and 5 seconds has elapsed.
+    // 5. Passed-in coinbaseSize differs from cached.
+    // 6. Passed-in coinbaseScript differs from cached.
+    if (pindexPrev != chainActive.Tip() || forceTemplateRecalc || // 1 & 2 above
+        (consensusParams.fPowAllowMinDifficultyBlocks && std::abs(GetTime() - nStart) > 30) || // 3 above
+        // 4 above
+        (mempool.GetTransactionsUpdated() != nTransactionsUpdatedLast && std::abs(GetTime() - nStart) > 5) ||
+        prevCoinbaseSize != coinbaseSize || !!prevCoinbaseScript != !!coinbaseScript || // 5 above
```

Comment:
nit: `bool(stuff)` is more readable than `!!stuff`

Reply:
Ok, will change.